backend/core/src/main/java/org/sonarsource/sonarlint/core/commons/SmartCancelableLoadingCache.java

```diff
@@ -0,0 +1,186 @@
+import java.util.function.BiFunction;
+import javax.annotation.Nullable;
+import org.sonarsource.sonarlint.core.commons.log.SonarLintLogger;
+import org.sonarsource.sonarlint.core.commons.progress.SonarLintCancelChecker;
+
+/**
+ * A cache with async computation of values, and supporting cancellation.
+ * "Smart" because when a computation is cancelled, it will return to the previous callers the result of the new computation.
+ */
+public class SmartCancelableLoadingCache<K, V> implements AutoCloseable {
+
+  private static final SonarLintLogger LOG = SonarLintLogger.get();
+
+  private final ExecutorService executorService;
+  private final String threadName;
+  private final BiFunction<K, SonarLintCancelChecker, V> valueComputer;
+  private final ConcurrentHashMap<K, ValueAndComputeFutures> cache = new ConcurrentHashMap<>();
+
+  private class ValueAndComputeFutures {
+    private final K key;
+    private CompletableFuture<V> valueFuture = new CompletableFuture<>();
+    private CompletableFuture<V> computeFuture;
+    private final ReentrantReadWriteLock lock = new ReentrantReadWriteLock();
+
+    public ValueAndComputeFutures(K key) {
+      this.key = key;
+      scheduleComputation();
+    }
+
+    private void scheduleComputation() {
+      lock.writeLock().lock();
+      if (computeFuture != null) {
+        computeFuture.cancel(false);
+        computeFuture = null;
+      }
+      // Even if we have canceled the previous computation, the previous valueFuture can still be completed later, so we will create a new one
+      // that will be returned to new callers. We will also try to complete the old valueFuture, in case the cancellation was successful.
+      var oldValueFuture = valueFuture;
+      valueFuture = new CompletableFuture<>();
+
+      var previousValue = oldValueFuture.getNow(null);
+
+      CompletableFuture<SonarLintCancelChecker> start = new CompletableFuture<>();
+      CompletableFuture<V> result = start.thenApplyAsync(cancelChecker -> valueComputer.apply(key, cancelChecker), executorService);
+      start.complete(new SonarLintCancelChecker(result));
+      result.whenCompleteAsync((newValue, error) -> {
+        lock.writeLock().lock();
+        try {
+          computeFuture = null;
+          if (error instanceof CancellationException) {
+            return;
+          }
+          try {
+            if (listener != null) {
+              listener.afterCachedValueRefreshed(key, previousValue, newValue);
+            }
+          } finally {
+            if (error != null) {
+              oldValueFuture.completeExceptionally(error);
+              valueFuture.completeExceptionally(error);
+            } else {
+              oldValueFuture.complete(newValue);
+              valueFuture.complete(newValue);
+            }
+          }
+        } finally {
+          lock.writeLock().unlock();
+        }
+      }, executorService);
+      computeFuture = result;
+      lock.writeLock().unlock();
```

Comment:
To be on the safe side, I would put all unlocking in a `finally` block, maybe nothing throws at the moment but later something could. This comment applies for all unlocks in this file